tests/jsunit/workspace_undo_redo_test.js

```diff
@@ -0,0 +1,386 @@
+ * @param {string} name The expected name of the variable.
+ * @param {string} type The expected type of the variable.
+ * @param {string} id The expected id of the variable.
+ */
+function undoRedoTest_checkVariableValues(name, type, id) {
+  var variable = workspace.getVariable(name);
+  assertNotUndefined(variable);
+  assertEquals(name, variable.name);
+  assertEquals(type, variable.type);
+  assertEquals(id, variable.getId());
+}
+
+/**
+ * Check if a variable names in specified top block.
+ * @param {number} blockIndex The index of the top block.
+ * @param {string} name The expected name of the variable in the block.
+ */
+function undoRedoTest_checkBlockVariableName(blockIndex, name) {
+  var block_var_name = workspace.topBlocks_[blockIndex].getVars()[0];
+  assertEquals(name, block_var_name);
+}
+
+function createTwoVarsEmptyType() {
+  workspace.createVariable('name1', '', 'id1');
+  workspace.createVariable('name2', '', 'id2');
+}
+
+function test_undoCreateVariable_Trivial() {
+  undoRedoTest_setUp();
+  workspace.createVariable('name1', 'type1', 'id1');
+  workspace.createVariable('name2', 'type2', 'id2');
+
+  workspace.undo();
+  undoRedoTest_checkVariableValues('name1', 'type1', 'id1');
+  assertNull(workspace.getVariableById('id2'));
+  workspace.undo();
+  assertNull(workspace.getVariableById('id1'));
+  assertNull(workspace.getVariableById('id2'));
+  undoRedoTest_tearDown();
+}
+
+function test_redoAndUndoCreateVariable_Trivial() {
+  undoRedoTest_setUp();
+  workspace.createVariable('name1', 'type1', 'id1');
+  workspace.createVariable('name2', 'type2', 'id2');
+
+  workspace.undo();
+  workspace.undo(true);
+
+  // Expect that variable 'id2' is recreated
+  undoRedoTest_checkVariableValues('name1', 'type1', 'id1');
+  undoRedoTest_checkVariableValues('name2', 'type2', 'id2');
+
+  workspace.undo();
+  workspace.undo();
+  workspace.undo(true);
+
+  // Expect that variable 'id1' is recreated
+  undoRedoTest_checkVariableValues('name1', 'type1', 'id1');
+  assertNull(workspace.getVariableById('id2'));
+  undoRedoTest_tearDown();
+}
+
+function test_undoDeleteVariable_NoBlocks() {
+  undoRedoTest_setUp();
+  workspace.createVariable('name1', 'type1', 'id1');
+  workspace.createVariable('name2', 'type2', 'id2');
+  workspace.deleteVariableById('id1');
+  workspace.deleteVariableById('id2');
+
+  workspace.undo();
+  assertNull(workspace.getVariableById('id1'));
+  undoRedoTest_checkVariableValues('name2', 'type2', 'id2');
+
+  workspace.undo();
+  undoRedoTest_checkVariableValues('name1', 'type1', 'id1');
+  undoRedoTest_checkVariableValues('name2', 'type2', 'id2');
+  undoRedoTest_tearDown();
+}
+
+function test_undoDeleteVariable_WithBlocks() {
+  undoRedoTest_setUpWithMockBlocks();
+  workspace.createVariable('name1', 'type1', 'id1');
+  workspace.createVariable('name2', 'type2', 'id2');
+  createMockBlock('name1');
+  createMockBlock('name2');
+  workspace.deleteVariableById('id1');
```

Comment:
It occurs to me that you should test what happens if you delete a variable by id twice--that is, test deleting a variable that doesn't exist.  Again, doesn't have  to be in this PR.